### ISSUE TITLE
Isomorphic scrum ?

### ISSUE THREAD

**Issue by DjebbZ** [roman01la/scrum#3]
Hello,

I want to start a discussion about using scrum in an isomorphic rum web app. We looked at the code and saw that a really small portion of the code uses javascript interop (mostly calls to `js/requestAnimationFrame`).

Why didn't you implement isomorphic scrum (with cljc) ? I know that scrum server-side brings almost no interest, but isomorphic does bring a lot.

What we understood from the code and Readme :

- initializing state by dispatching a scrum event with `broadcast-sync!` already works both server and client side
- `subscription` just needs a Clojure implementation that does a `get-in` with the given path in the state. Need to think about multithreading though... Maybe it's a problem.

I hope you engage in this discussion and that we find a way to do it. Thanks in advance !

**Comment by roman01la** [comment]
Hi @DjebbZ. I just never needed it 🙃
I'm excited to accept a PR which would bring Clojure support.
JS interop could be hidden behind reader condition. But I'm not sure about everything else.

**Comment by DjebbZ** [comment]
I forgot to say reader macro could help for interop, sure.

Forget about what I said about `subscription` and `get-in`, it was part of our plan to workaround non-isomorphic scrum and make it work with isomorphic rum.

I hope I find some time to make a PR. Hopefully this weekend or next week.

**Comment by roman01la** [comment]
Do you even need subscriptions on the server? This is a part of reactive UI flow, not sure how this is useful on the backend.

**Comment by DjebbZ** [comment]
Of course subscriptions and reactions are useless in the server, but if
scrum is written in cljc and made isomorphic it basically means that as a
user of scrum I don't need to bother with splitting code between client and
server for the UI.

**Comment by roman01la** [comment]
@DjebbZ FYI, I just published experimental SSR support in Scrum as `2.0.0-SNAPSHOT`, there's also usage example in [scrum-ssr-example](https://github.com/roman01la/scrum-ssr-example) repo

**Comment by DjebbZ** [comment]
Wow so fast ! No time to play with it this weekend, but be sure that we'll
try next week. I'll keep you updated.

**Comment by roman01la** [comment]
@DjebbZ Another update. It includes some changes to introduce an ability to retrieve fetched data during server rendering so it can be then serialized and rendered into HTML for web app to pick it up when it's loaded.

In short:
```clj
;; declare subscription resolvers
(defn make-resolver [db]
  {[:counter] #(query-db db)})

;; create reconciler by passing it resolvers hash
(def reconciler (-> db make-resolver scrum/reconciler))

;; render application
(def html (-> reconciler AppComponent rum/render-html))

;; retrieve app state populated during render
(def state @(:state reconciler))

;; render entire HTML document
(render-page html state)
```

I'll update example repo later

**Comment by DjebbZ** [comment]
Studying the ssr-example right now. Thanks for the hard work ! From afar it looks you did everything already.

**Comment by roman01la** [comment]
@DjebbZ I just updated the example with up to date Scrum implementation

**Comment by DjebbZ** [comment]
Nice !

**Comment by DjebbZ** [comment]
Regarding parallel/sequential resolving :

From the point of view of a scrum user having resolvers server-side and controllers client-side feel a bit weird. But overfetching is really something bad. I'm currently implementing ssr right now (client-side done, server-side in progress), I'll keep you updated about how it goes.

**Comment by roman01la** [comment]
I want to add that I'm not really sure if front-end architecture should match back-end when speaking of controllers. I'd like to have more flexibility here 🤔

**Comment by DjebbZ** [comment]
Agree. The wording "resolver" is still nice.

**Comment by DjebbZ** [comment]
I have an issue right now. I refactored a small part of our app (based an a global atom) with scrum with ssr but I just can't compile it... It's completely stuck at the cljs compilation phase. Do you have the same problem ?

**Comment by roman01la** [comment]
@DjebbZ Works fine for me. Could you provide more useful info? Maybe `project.clj` file. Also try to add `:verbose true` to compiler options to see detailed output during compilation.

**Comment by DjebbZ** [comment]
I use boot not leiningen, but I tried to pass `:verbose true` to compiler options and it seems to get stuck on some file of mine that uses scrum... Maybe there's some weird interaction between the two. I noticed it only happens when compile the clojurescript side, the clojure side is fine.

**Comment by roman01la** [comment]
Just published an update to `2.1.0-SNAPSHOT` and up to date documentation, see https://github.com/roman01la/scrum#server-side-rendering

**Comment by DjebbZ** [comment]
It works fine. Closing.
